Packs/IBMSecurityVerify/ModelingRules/IBMSecurityVerify/IBMSecurityVerify.xif

```diff
@@ -0,0 +1,139 @@
+    xdm.source.ipv4 = if(incidr(data -> origin, "0.0.0.0/0"), data -> origin),
+    xdm.source.ipv6 = src_ipv6,
+    xdm.source.user_agent = data -> devicetype,
+    xdm.observer.name = if(tenantname != null AND servicename != null, concat(tenantname, " - ", servicename), tenantname),
+    xdm.observer.unique_identifier = tenantid;
+
+[MODEL: dataset=ibm_security_verify_raw]
+/*
+Mapping of 'management' event types.
+*/
+filter event_type = "management"
+| alter
+    src_user1 = arrayindex(regextract(data -> performedby_username, "(.+)\@") ,0), // The username can sometimes appear as a username and other times as an email.
+    src_user2 = arrayindex(regextract(data -> username, "(.+)\@") ,0),
+    mod = if(data -> modified = "[]", null, data -> modified)
+| alter
+    xdm.event.operation_sub_type = coalesce(data -> dict_op, data -> action),
+    xdm.target.application.name = data -> applicationname,
+    xdm.event.description = if(data -> cause != null AND data -> modified != null, concat(rtrim(data -> cause, "."), ". modified properties: ", mod), coalesce(data -> cause, data -> modified)), // Checks if 'cause' and 'modified' are both populated. If so, appends the cause with a list of modified properties. If not, returns whichever field is available.
+    xdm.event.original_event_type = data -> messageid,
+    xdm.auth.mfa.method = data -> mfamethod,
+    xdm.source.user.identifier = if(data -> performedby IN("System", "shared_apikey"), null, data -> performedby),
+    xdm.source.application.name = coalesce(data -> performedby_clientname, if(data -> performedby IN("System", "shared_apikey"), data -> performedby)),
+    xdm.source.user.username = coalesce(src_user2, data -> username, src_user1, data -> performedby_username), //2 possible username fields, performedby_username and username.
+    xdm.source.user.upn = if(src_user2 != null, data -> username ,src_user1 != null, data -> performedby_username),
+    xdm.target.resource.id = data -> entitlementid,
+    xdm.target.resource.name = data -> entitlementname,
+    xdm.target.resource.sub_type = data -> entitlementsubtype,
+    xdm.target.resource.type = if(data -> entitlementtype != null AND data -> entitlementcategory != null, concat(data -> entitlementtype, " - ", data -> entitlementcategory), coalesce(data -> entitlementtype, data -> entitlementcategory, data -> resource )), // The "type" field represents the role, while the "category" field specifies the role type, e.g., "role - admin."
+    xdm.source.user.groups = to_json_string(data -> entitlements) -> []
+| call generic_fields_mapping;
+
+/*
+Mapping of 'sso' event types.
+*/
+filter event_type = "sso"
+| alter
+    src_user = arrayindex(regextract(data -> username, "(.+)\@") ,0) // The username can sometimes appear as a username and other times as an email.
+| alter
```

Comment:
For the the relevant events for 102 story, in our case it's "**sso**" and "**authentication**" events,  we need to verify the following:
1. xdm.event.type = "authentication"
2. xdm.event.tags = arraycreate(XDM_CONST.EVENT_TAG_AUTHENTICATION)
3. Have 5 tuples (even zero if not exists)
4. All of the following fields contain data:
- xdm.source.ipv4
- xdm.source.user.upn
- xdm.event.original_event_type
- xdm.event.outcome
- xdm.event.outcome_reason
- xdm.event.operation

p.s 
you can talk with Eido Epstain as he very familiar with it.

Reply:
A couple of questions:

1. Regarding the 5 tuples, there is no information about whether it's TCP/UDP. Please specify the field from the XDM schema where you want this to be populated. I found only application protocol-related fields.
2. In `xdm.event.original_event_type`, it's currently populated with `messageid`, a unique ID number for the message in data.cause. This ID refers to a specific event subtype, in this case, authentication. 
3. What should be populated in `xdm.event.outcome_reason`?
4. Since `xdm.event.operation` is an ENUM and we don't have all the possible event subtypes documented in the IBM documentation, I've populated `xdm.event.operation_sub_type` for the 'authentication' event type. However, for 'sso' events, we don't have any additional information that can be extracted from the logs.

Reply:
@yassorpanw 

1. xdm.event.tags - Verified
2. xdm.source.user.upn - Verified
3. xdm.event.type = "authentication" - Verified
4. xdm.event.original_event_type - Verified
5. xdm.event.outcome - Verified
6. xdm.event.operation (XDM_CONST.OPERATION_TYPE_AUTH_LOGIN) - Added
7. xdm.event.outcome_reason - Waiting for more information / assistance from IBM.
8. 5 Tuples - Verified and added except for **xdm.network.ip_protocol** as discussed. I also didn't see that field populated in either Duo or Okta.

To summarize, there are 2 open things:

1. xdm.event.outcome_reason
2. 5 Tuples - Your response about what should we do regarding the **xdm.network.ip_protocol**.